### PR TITLE
Refactor/add bit check to sign expr

### PR DESCRIPTION
**Pull request by stuarttimwhite**
Please be sure to look over the pull request guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md#submit-pr.

# Please go through the following checklist
- [ ] The PR title and commit messages adhere to guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md. In particular `!` is used if and only if at least one breaking change has been introduced.
- [ ] I have run the ci check script with `source scripts/run_ci_checks.sh`.

# Rationale for this change

The code in sign_expr could verify that values use less than or equal to a given number of bits, but it does not currently. This PR adds that functionality.

# What changes are included in this PR?

Adding a parameter num_bits_allowed to verify_bit_decomposition and verfify_sign_expr, along with the corresponding functionality.

# Are these changes tested?
Yes.

### CODE REVIEW

File: crates/proof-of-sql/src/sql/proof_gadgets/sign_expr.rs [hunk 1]
```diff
@@ -159,7 +162,13 @@ fn verify_bit_decomposition<S: ScalarExt>(
             rhs += S::from_wrapping(mult) * bit_eval;
         }
     }
-    (rhs == expr_eval)
+    let bits_that_must_match_inverse_lead_bit = U256::MAX
+        .shl(num_bits_allowed.unwrap_or(S::MAX_BITS).min(S::MAX_BITS) - 1)
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_gadgets/sign_expr.rs:166]: NIT: The `min` is a bit concerning here. I'd rather make this return an error if the allowed bits is larger than MAX_BITS.

File: crates/proof-of-sql/src/base/scalar/mont_scalar.rs [hunk 2]
```diff
@@ -444,6 +444,14 @@ where
...
         ])
     };
+    #[allow(clippy::cast_possible_truncation)]
+    const MAX_BITS: u8 = {
+        assert!(
+            T::MODULUS.0[3].leading_zeros() < 64,
+            "modulus expected to be larger than 1 << (64*3)"
+        );
+        255 - T::MODULUS.0[3].leading_zeros() as u8
+    };
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/scalar/mont_scalar.rs:454]: Add some tests here.

File: crates/proof-of-sql/src/sql/proof_gadgets/sign_expr.rs [hunk 3]
```diff
@@ -159,7 +162,13 @@ fn verify_bit_decomposition<S: ScalarExt>(
...
+        ^ U256::ONE.shl(255);
+    let is_eval_too_many_bits = bits_that_must_match_inverse_lead_bit
+        & dist.leading_bit_inverse_mask()
+        == bits_that_must_match_inverse_lead_bit;
+    (rhs == expr_eval && is_eval_too_many_bits)
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_gadgets/sign_expr.rs:171]: It would be nice to add some tests for the large edge cases.

**Review by JayWhite2357 — APPROVED** [review]

**Review by JayWhite2357 — APPROVED** [review]
